File: services/code-service/internal/tracing/tracing.go
// Package tracing initialise OpenTelemetry pour exporter les traces vers Tempo (OTLP/gRPC).
package tracing

import (
	"context"
	"net/http"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.24.0"
)

// Init configure le TracerProvider global avec un exporteur OTLP/gRPC vers endpoint (ex: "tempo:4317").
// Retourne une fonction shutdown à appeler à l'arrêt du service.
func Init(ctx context.Context, serviceName, endpoint string) (func(context.Context) error, error) {
	exporter, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(endpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, err
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(serviceName),
		),
	)
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter, sdktrace.WithBatchTimeout(5*time.Second)),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.AlwaysSample()),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

// SkipHealthAndMetrics indique à otelgin de ne pas tracer les endpoints
// de healthcheck et de scraping Prometheus, trop fréquents pour être utiles dans Tempo.
func SkipHealthAndMetrics(r *http.Request) bool {
	switch r.URL.Path {
	case "/health", "/healthz", "/metrics":
		return false
	}
	return true
}

File: services/ansible-service/internal/tracing/tracing.py
"""Initialise OpenTelemetry pour exporter les traces vers Tempo (OTLP/gRPC)."""
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource, SERVICE_NAME
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor


def init_provider(service_name: str, endpoint: str) -> None:
    """Configure le TracerProvider global avec un exporteur OTLP/gRPC vers endpoint (ex: "tempo:4317")."""
    resource = Resource(attributes={SERVICE_NAME: service_name})
    provider = TracerProvider(resource=resource)
    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)


def instrument_app(app) -> None:
    """Instrumente l'application FastAPI (doit être appelé avant le démarrage de l'app)."""
    # Exclure les endpoints de healthcheck et de scraping Prometheus, trop fréquents
    # pour être utiles dans Tempo.
    FastAPIInstrumentor.instrument_app(app, excluded_urls="/health,/metrics")
